File: homework/Truong_Duong/lesson_1/303_Range_sum_query_Immutable.py
# Solution for https://leetcode.com/problems/range-sum-query-immutable/

class NumArray:

    def __init__(self, nums: List[int]):
        self.data = [0] * len(nums)

        self.data[0] = nums[0]
        for i in range(1, len(nums)):
            self.data[i] = self.data[i-1] + nums[i]


    def sumRange(self, left: int, right: int) -> int:
        return self.data[right] if left == 0 else self.data[right] - self.data[left-1]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)

File: homework/Thu_Pham/lesson_1/303-range-sum-query-immutable.js
//Solution for https://leetcode.com/problems/range-sum-query-immutable/

/**
 * @param {number[]} nums
 */
var NumArray = function (nums) {
  this.nums = nums;
};

/**
 * @param {number} left
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function (left, right) {
  const prefix = [];
  prefix.push(this.nums[0]);
  for (let i = 1; i < this.nums.length; i++) { // O(n)
    prefix.push(this.nums[i] + prefix[i - 1]);
  }
  return left === 0 ? prefix[right] : prefix[right] - prefix[left - 1]; //O(1)
};

//runtime O(n) with n is the length of the array
//space O(n) with n is the length of the array 

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */
